src/pluggy/callers.py

```diff
@@ -109,6 +109,7 @@ def _multicall(hook_impls, caller_kwargs, firstresult=False):
                         next(gen)  # first yield
                         teardowns.append(gen)
                     except StopIteration:
+                        gen.close()
```

Comment:
Can you describe why this is needed?

(I think it was discussed before but I'm still not sure).

BTW, I think it will be nice to move the `gen = hook_impl.function(*args)` line to before the `try`, and the `teardowns.append(gen)` to after the `try`, because it's good to keep the scope of `try` as focused as possible.

Reply:
This code is executed for pathological hooks only so I do not think it is really important.

A real step to improve diagnostics is to add a check that hookwrapper is a generator on the initialization step before attempts to run hooks. However `gen.close()` is related to bad generator, so it is another case.

I do not mind to move `gen =` line above `try`.

Really defensive style would require `gen.close()` in response to any other exception, including improbable failure of `teardowns.append()`.

`close()` was added here in response to the comments of @goodboy.

https://github.com/pytest-dev/pluggy/pull/257#issuecomment-619628958:
> In summary, I think we should make 2 changes:
> - `generator.close()` should always be called

I missed that other related comments have been edited: https://github.com/pytest-dev/pluggy/pull/257#issuecomment-619423465 https://github.com/pytest-dev/pluggy/pull/257#issuecomment-619424727

I do not have strong opinion concerning `close` here. It makes no harm. In principle it might allow interpreter implementation to release resources earlier but I am not sure. Taking into account https://github.com/pytest-dev/pluggy/pull/257#pullrequestreview-421601668 I have realized just now that there are more objections than arguments in support of the change. Actually I was waiting for more opinions from developers.

Reply:
> A real step to improve diagnostics is to add a check that hookwrapper is a generator on the initialization step before attempts to run hooks.

I think that's a good idea, sent #282 which adds this validation.

> Really defensive style would require gen.close() in response to any other exception, including improbable failure of teardowns.append().

I think we can assume that `teardowns.append` does not fail.

> I do not have strong opinion concerning close here. It makes no harm. In principle it might allow interpreter implementation to release resources earlier but I am not sure

The docs [say](https://docs.python.org/3.8/reference/expressions.html#generator.close)

    close() does nothing if the generator has already exited due to an exception or normal exit.

This is the case here, so I think that that `close` is redundant, unless I'm missing something.

Similarly to the defensive approach you suggested, IIUC, if the `next(gen)` raises it means the `close()` will do nothing.

Reply:
Agreed with @bluetech -- the extra `gen.close()` isn't doing anything here.